### PR TITLE
Small refactoring of sample_gp

### PR DESCRIPTION
**Pull request by denadai2**
I fixed a potential error on the correction variable, and refactored a bit the code

### CODE REVIEW

File: pymc3/gp/gp.py [hunk 1]
```diff
@@ -106,18 +108,18 @@ def sample_gp(trace, gp, X_values, samples=None, obs_noise=True, model=None, ran
     -------
     Array of samples from posterior GP evaluated at Z.
     """
-    model = modelcontext(model)
-
     if samples is None:
         samples = len(trace)
 
+    if model is None:
```
Comment (bwengals) [inline pymc3/gp/gp.py]: I think there is no need for None check, `modelcontext` [does this already](https://github.com/pymc-devs/pymc3/blob/master/pymc3/model.py#L144)

File: pymc3/gp/gp.py [hunk 2]
```diff
@@ -134,11 +136,13 @@ def sample_gp(trace, gp, X_values, samples=None, obs_noise=True, model=None, ran
...
 
+    S_xz_S_inv = tt.dot(S_xz.T, S_inv)
     # Posterior mean
-    m_post = tt.dot(tt.dot(S_xz.T, S_inv), Y)
+    m_post = tt.dot(S_xz_S_inv, Y)
     # Posterior covariance
-    S_post = S_zz - tt.dot(tt.dot(S_xz.T, S_inv), S_xz)
+    S_post = S_zz - tt.dot(S_xz_S_inv, S_xz)
 
+    correction = 0
```
Comment (bwengals) [inline pymc3/gp/gp.py:144]: Good catch!  This should have been caught in tests (which is my bad!)

File: pymc3/gp/gp.py [hunk 3]
```diff
@@ -134,11 +136,13 @@ def sample_gp(trace, gp, X_values, samples=None, obs_noise=True, model=None, ran
     else:
         S_inv = matrix_inverse(K(X))
 
+    S_xz_S_inv = tt.dot(S_xz.T, S_inv)
```
Comment (bwengals) [inline pymc3/gp/gp.py:138]: I think theano takes care of optimizations like this

**Review by bwengals — COMMENTED** [review]
I think this should be merged asap, given the bug fix (line 145)

File: pymc3/gp/gp.py (hunk 3 above)
Comment (denadai2) [reply pymc3/gp/gp.py:138]: I was refactoring, so I did this :P no harm, less code

File: pymc3/gp/gp.py (hunk 1 above)
Comment (denadai2) [reply pymc3/gp/gp.py]: ok so I remove it from here and here https://github.com/pymc-devs/pymc3/blob/master/pymc3/sampling.py#L529?

**Comment by junpenglao** [comment]
@denadai2 Could you please amend the changes above?

**Comment by denadai2** [comment]
done :) I also removed the check from sample_ppc

**Comment by junpenglao** [comment]
Thanks @denadai2 !

**Comment by denadai2** [comment]
np

**Pull request merged**